### ISSUE TITLE
ADD --from

### ISSUE THREAD

**Issue by joshwget** [moby/moby#33811]
Any plans to support `--from` for `ADD` as it is for `COPY`?

```
FROM <>
ADD --from=0 /<>.tar.gz /<>/<>
```

I could see this being a pretty powerful feature and so far I haven't been able to think of a downside to supporting it. If this ends up being something agreed on I'd be interesting in contributing this feature.

**Comment by justincormack** [comment]
Why can't you just untar in an intermediate image and then COPY the output?

**Comment by joshwget** [comment]
AFAIK `COPY` doesn't work for nested directories in the same way as `ADD`. If `COPY` worked like `cp -r` then I wouldn't see much value in supporting this.

**Comment by justincormack** [comment]
There are some issues about fixing copy. Now we have flags I would love to
have one that makes it non broken. Supporting ADD is worse. Or just switch
to CP with the correct semantics.

**Comment by joshwget** [comment]
If you mean something like the following that would certainly accomplish the same goal for me. Supporting `ADD` was just the first way I thought to do it.

```
FROM <>
COPY/CP --from=0 -r /<> /<>/<>
```

If adding flags such as these to `COPY` already has an open issue we can probably close this. Once the flags are added I'd think they'd just work when used with `--from`.

**Comment by cpuguy83** [comment]
`COPY` should be identical to `ADD` except it doesn't untar automatically (e.g. `ADD foo.tar /data`) and it doesn't fetch external resources (e.g. `ADD http://some/url/myFile /data`)

**Comment by joshwget** [comment]
@cpuguy83 Correct, I was planning to use the untar functionality of `ADD` as a replacement for recursive copying with `COPY`.

**Comment by thaJeztah** [comment]
Instead of overloading `COPY`, I'd be in favour of https://github.com/moby/moby/issues/32507 (`RUN --mount`), which allows you to use any shell command (including `cp`)

**Comment by joshwget** [comment]
@thaJeztah I think that would work except in the case of `FROM scratch` images.

**Comment by olemis** [comment]
@thaJeztah I'd be in favor of adding `--from` to `ADD` command over `RUN --mount` and I'll explain with a real-life example related to cross-compilation. Please do not interpret my comment as being against #32570 . That's another subject.

We are building images for ARM using a [single Dockerfile](https://github.com/skycoin/skycoin/blob/develop/docker/images/mainnet/Dockerfile) (see [skycoin/skycoin build hook](https://github.com/skycoin/skycoin/blob/develop/docker/images/mainnet/hooks/build) if interested on the details) . Recently we were trying to reduce the number of layers by creating an archive in `build` intermediate image and then copying its contents to production image  .

The straightforward approach would be `COPY --from build ...` followed by `RUN tar x...` . This succeeds for `amd64` architectures. Nonetheless while building ARM images `RUN`ning anything would fail with message `standard_init_linux.go:190: exec user process caused "exec format error"` , hence forcing the use of `qemu` , and other complex approaches (which I'll not explain for the sake of being concise) .

By supporting `ADD --from ` the very common situation of copying an archive built in an intermediate image and extracting its files in target production image would be firstly , intuitive due to its similarity to `COPY --from` , but more important, will be performed by Docker itself , thereby this will work while building images for different architectures .

p.s. this also answers @justincormack question above `Why can't you just untar in an intermediate image and then COPY the output?`.

**Comment by olemis** [comment]
Another approach consists in `--squash`ing layers in `build` hook , but that does not seem to be supported (yet?) in Docker Cloud builds.

**Comment by cpuguy83** [comment]
Little bit of a work-around your case, but it's also not bad...

```
FROM scratch AS data
ADD data.tar /data

FROM foo
COPY --from=data ...
```

**Comment by olemis** [comment]
@cpuguy83 what you mention is similar to [what we are doing now in develop](https://github.com/skycoin/skycoin/blob/develop/docker/images/mainnet/Dockerfile#L42-L52) . Nonetheless , since destination folders are different this leads to an [image with four different layers](https://microbadger.com/images/skycoin/skycoin:release-v0.24.1) . We are trying to bring this down to a single layer since image as a whole is about 30 MB in size.

I am not sure we can use `ADD` the way you mention due to the fact that it's created in another intermediate image , not in the build context (as you suggest) . I might have misunderstood your precious comment though .

**Comment by cpuguy83** [comment]
@olemis 

Why not extract the tarball in the other image and copy the dir over in that case?

**Comment by olemis** [comment]
Something like that's what we've actually done in skycoin/skycoin#1903 , so I'm kind of changing my mind about this issue . This feature now looks like redundant to me.

**Comment by JtMotoX** [comment]
I know this is an old post but it is still marked as open so adding a solution here if anyone else stumbles across this.

If you have buildkit enabled (`DOCKER_BUILDKIT=1`), you can mount the file within the layer and extract it without actually copying it to your image.  When you mount a file or directory, it does not add any size to the layer. The only size taken within this layer is the resulting extracted directory.

```dockerfile
RUN --mount=type=bind,from=<>,source=/foo/bar.tar.gz,target=/foo/bar.tar.gz \
	tar -xvf /foo/bar.tar.gz
```
